node/network/availability-recovery/src/lib.rs

```diff
@@ -0,0 +1,843 @@
+		session_index,
+		ctx,
+	).await?.await.map_err(error::Error::CanceledSessionInfo)??;
+
+	match session_info {
+		Some(session_info) => {
+			launch_interaction(
+				state,
+				ctx,
+				session_index,
+				session_info,
+				receipt,
+				response_sender,
+			).await
+		}
+		None => {
+			tracing::warn!(
+				target: LOG_TARGET,
+				"SessionInfo is `None` at {}", state.live_block_hash,
+			);
+			response_sender
+				.send(Err(RecoveryError::Unavailable))
+				.map_err(|_| error::Error::CanceledResponseSender)?;
+			Ok(())
+		}
+	}
+}
+
+/// Queries a chunk from av-store.
+#[tracing::instrument(level = "trace", skip(ctx), fields(subsystem = LOG_TARGET))]
+async fn query_chunk(
+	ctx: &mut impl SubsystemContext<Message = AvailabilityRecoveryMessage>,
+	candidate_hash: CandidateHash,
+	validator_index: ValidatorIndex,
+) -> error::Result<Option<ErasureChunk>> {
+	let (tx, rx) = oneshot::channel();
+	ctx.send_message(AllMessages::AvailabilityStore(
+		AvailabilityStoreMessage::QueryChunk(candidate_hash, validator_index, tx),
+	)).await;
+
+	Ok(rx.await.map_err(error::Error::CanceledQueryChunk)?)
+}
+
+/// Handles message from interaction.
+#[tracing::instrument(level = "trace", skip(ctx, state), fields(subsystem = LOG_TARGET))]
+async fn handle_from_interaction(
+	state: &mut State,
+	ctx: &mut impl SubsystemContext<Message = AvailabilityRecoveryMessage>,
+	from_interaction: FromInteraction,
+) -> error::Result<()> {
+	match from_interaction {
+		FromInteraction::Concluded(candidate_hash, result) => {
+			// Load the entry from the interactions map.
+			// It should always exist, if not for logic errors.
+			if let Some(interaction) = state.interactions.remove(&candidate_hash) {
+				// Send the result to each member of awaiting.
+				for awaiting in interaction.awaiting {
+					if let Err(_) = awaiting.send(result.clone()) {
+						tracing::debug!(
+							target: LOG_TARGET,
+							"An awaiting side of the interaction has been canceled",
+						);
+					}
+				}
+			} else {
+				tracing::warn!(
+					target: LOG_TARGET,
+					"Interaction under candidate hash {} is missing",
+					candidate_hash,
+				);
+			}
+
+			state.availability_lru.put(candidate_hash, result);
+		}
+		FromInteraction::MakeRequest(id, candidate_hash, relay_parent, validator_index, response) => {
+			// Take the validators we are already connecting to and merge them with the
+			// newly requested one to create a new connection request.
+			let new_discovering_validators_set = state.discovering_validators.keys()
+				.cloned()
+				.chain(std::iter::once(id.clone()))
+				.collect();
+
+			// Issue a NetworkBridgeMessage::ConnectToValidators.
+			// Add the stream of connected validator events to state.connecting_validators.
+			match validator_discovery::connect_to_validators(
+				ctx,
+				relay_parent,
+				new_discovering_validators_set,
+			).await {
+				Ok(new_connection_request) => {
+					state.connection_requests.put(relay_parent, new_connection_request);
```

Comment:
The doc on `connection_requests` says that this will drop any other request from the `relay_parent`. I think this API isn't really suitable to our needs. Maybe we should just use `ConnectToValidators` directly and save ourselves the headache.

Reply:
The main question I have is about ownership of the receiver, which unfortunately I overlooked when writing the guide page. I'd guess that we want the receiver to drop as soon as we either fulfill the request or time out. At the same time, we want the main subsystem state to be the one polling the receiver. I'm sure there's a workaround, but maybe kind of ugly and involves more channels.